nav2_graceful_motion_controller/src/graceful_motion_controller.cpp

```diff
@@ -0,0 +1,358 @@
+  bool reversing = false;
+  if (params_->allow_backward && motion_target.pose.position.x < 0.0) {
+    reversing = true;
+    motion_target.pose.orientation = nav2_util::geometry_utils::orientationAroundZAxis(
+      tf2::getYaw(motion_target.pose.orientation) + M_PI);
+  }
+
+  // Compute velocity command:
+  // 1. Check if we are close enough to the goal to do a final rotation in place
+  // 2. Check if we must do a rotation in place before moving
+  // 3. Calculate the new velocity command using the smooth control law
+  geometry_msgs::msg::TwistStamped cmd_vel;
+  cmd_vel.header = pose.header;
+  if (params_->final_rotation && (dist_to_goal < goal_dist_tolerance_ || goal_reached_)) {
+    goal_reached_ = true;
+    double angle_to_goal = tf2::getYaw(transformed_plan.poses.back().pose.orientation);
+    cmd_vel.twist = rotateToTarget(angle_to_goal);
+  } else if (params_->initial_rotation && // NOLINT
+    fabs(angle_to_target) > params_->initial_rotation_min_angle)
+  {
+    cmd_vel.twist = rotateToTarget(angle_to_target);
+  } else {
+    cmd_vel.twist = control_law_->calculateRegularVelocity(motion_target.pose, reversing);
+  }
+
+  // Transform local frame to global frame to use in collision checking
+  geometry_msgs::msg::TransformStamped costmap_transform;
+  try {
+    costmap_transform = tf_buffer_->lookupTransform(
+      costmap_ros_->getGlobalFrameID(), costmap_ros_->getBaseFrameID(),
+      tf2::TimePointZero);
+  } catch (tf2::TransformException & ex) {
+    RCLCPP_ERROR(
+      logger_, "Could not transform %s to %s: %s",
+      costmap_ros_->getBaseFrameID().c_str(), costmap_ros_->getGlobalFrameID().c_str(),
+      ex.what());
+    return cmd_vel;
+  }
+
+  // Generate and publish local plan for debugging / visualization
+  nav_msgs::msg::Path local_plan;
+  if (!simulateTrajectory(pose, motion_target, costmap_transform, local_plan, reversing)) {
+    throw nav2_core::NoValidControl("Collision detected in the trajectory");
+  }
+  local_plan.header = transformed_plan.header;
+  local_plan_pub_->publish(local_plan);
+
+  return cmd_vel;
+}
+
+void GracefulMotionController::setPlan(const nav_msgs::msg::Path & path)
+{
+  path_handler_->setPlan(path);
+  goal_reached_ = false;
+}
+
+void GracefulMotionController::setSpeedLimit(
+  const double & speed_limit, const bool & percentage)
+{
+  std::lock_guard<std::mutex> param_lock(param_handler_->getMutex());
+
+  if (percentage) {
+    // Speed limit is expressed in % from maximum speed of robot
+    params_->v_linear_max = std::max(
+      params_->v_linear_max_initial * speed_limit / 100.0,
+      params_->v_linear_min);
+  } else {
+    // Speed limit is expressed in m/s
+    params_->v_linear_max = std::max(speed_limit, params_->v_linear_min);
+  }
+
+  // Limit the angular velocity to be proportional to the linear velocity
+  params_->v_angular_max = params_->v_linear_max * params_->rotation_scaling_factor;
```

Comment:
Huh? I think you need to rethink this.

Reply:
Why would either of these values appear in this expression? It should be something like the proportion of max linear velocity reduction times the original max angular velocity to scale it down proportionately to the requested linear velocity reduction. Try checking out other implementations